_posts/2022-10-30-multi-column-sorts-in-arrow-rust-part-2.md

```diff
@@ -0,0 +1,243 @@
+
+```
+       ┌──┬──┬──┬──┐       ┌──┬──┬──┬──┐       ┌──┬──┬──┬──┬──┐
+    5  │05│00│00│00│       │85│00│00│00│       │01│00│00│00│85│
+       └──┴──┴──┴──┘       └──┴──┴──┴──┘       └──┴──┴──┴──┴──┘
+       ┌──┬──┬──┬──┐       ┌──┬──┬──┬──┐       ┌──┬──┬──┬──┬──┐
+   -5  │FB│FF│FF│FF│       │7B│FF│FF│FF│       │01│FF│FF│FF│FF│
+       └──┴──┴──┴──┘       └──┴──┴──┴──┘       └──┴──┴──┴──┴──┘
+
+ Value  32-bit (4 bytes)    High bit flipped      Row Format
+         Little Endian
+```
+
+### Floating Point
+
+Floating point values can be ordered according to the [IEEE 754 totalOrder predicate](https://en.wikipedia.org/wiki/IEEE_754#Total-ordering_predicate) (implemented in Rust by [f32::total_cmp](https://doc.rust-lang.org/std/primitive.f32.html#method.total_cmp)). This ordering interprets the bytes of the floating point value as the correspondingly sized, signed, little-endian integer, flipping all the bits except the sign bit in the case of negatives.
+
+Floating point values are therefore encoded to row format by converting them to the appropriate sized signed integer representation, and then using the same encoding for signed integers described in the previous section.
+
+### Byte Arrays (Including Strings)
+
+Unlike primitive types above, byte arrays are variable length. For short strings, such as `state` in our example above, it is possible to pad all values to the length of the longest one with some fixed value such as `0x00` and produce a fixed length row. This is the approach described in the DuckDB blog for encoding `c_birth_country`.
+
+However, often values in string columns differ substantially in length or the maximum length is not known at the start of execution, making it inadvisable and/or impractical to pad the strings to a fixed length. The Rust Arrow row format therefore uses a variable length encoding.
+
+We need an encoding that unambiguously terminates the end of the byte array. This not only permits recovering the original value from the row format, but ensures that bytes of a longer byte array are not compared against bytes from a different column when compared against a row containing a shorter byte array.
+
+A null byte array is encoded as a single `0x00` byte. Similarly, an empty byte array is encoded as a single `0x01` byte.
+
+To encode a non-null, non-empty array, first a single `0x02` byte  is written. Then the array is written in 32-byte blocks, with each block followed by a `0xFF` byte as a continuation token. The final block is padded to 32-bytes with `0x00`, and is then followed by the unpadded length of this final block as a single byte.
```

Comment:
```suggestion
To encode a non-null, non-empty array, first a single `0x02` byte  is written. Then the array is written in 32-byte blocks, with each complete block followed by a `0xFF` byte as a continuation token. The final block is padded to 32-bytes with `0x00`, and is then followed by the unpadded length of this final block as a single byte in place of a continuation token
```
I think this makes it clearer that the continuation token is for when the blocks continue :smile: